Changes to add CMD for runtime images

jhuttana opened the pull request:
***Could you please review the changes?***


# Case-1 : ubi8/openjdk-8-runtime
```
$ podman inspect --format "{{.Config.Cmd}}" localhost/ubi8/openjdk-8-runtime:latest
    [/opt/jboss/container/java/run]

$ podman run --rm -ti localhost/ubi8/openjdk-8-runtime:latest
    Starting the Java application using /opt/jboss/container/java/run/run-java.sh ...
    This is the Red Hat UBI8 OpenJDK 1.8.0 Runtime container.

$ podman run --rm -ti localhost/ubi8/openjdk-8-runtime:latest rpm -qa | grep "prometheus"
$ podman run --rm -ti localhost/ubi8/openjdk-8-runtime:latest rpm -qa | grep "maven"
$ podman run --rm -ti localhost/ubi8/openjdk-8-runtime:latest rpm -qa | grep "java"
    java-1.8.0-openjdk-headless-1.8.0.282.b08-2.el8_3.x86_64
    tzdata-java-2021a-1.el8.noarch
    javapackages-filesystem-5.3.0-1.module+el8+2447+6f56d9a6.noarch
```

# Case-2: ubi8/openjdk-11-runtime
```
$ podman inspect --format "{{.Config.Cmd}}" localhost/ubi8/openjdk-11-runtime:latest
    [/opt/jboss/container/java/run]

$ podman run --rm -ti localhost/ubi8/openjdk-11-runtime:latest
    Starting the Java application using /opt/jboss/container/java/run/run-java.sh ...
    This is the Red Hat UBI8 OpenJDK 11 Runtime container.

$ podman run --rm -ti localhost/ubi8/openjdk-11-runtime:latest rpm -qa | grep "maven"
$ podman run --rm -ti localhost/ubi8/openjdk-11-runtime:latest rpm -qa | grep "prometheus"
$ podman run --rm -ti localhost/ubi8/openjdk-11-runtime:latest rpm -qa | grep "java"
    javapackages-filesystem-5.3.0-1.module+el8+2447+6f56d9a6.noarch
    java-11-openjdk-headless-11.0.10.0.9-4.el8_3.x86_64
    tzdata-java-2021a-1.el8.noarch
```
Also, 
```
$ podman images --filter dangling=false
REPOSITORY                               TAG  		  SIZE
localhost/ubi8/openjdk-8-runtime         1.3              314 MB
localhost/ubi8/openjdk-8-runtime         latest           314 MB
localhost/ubi8/openjdk-11-runtime        latest           370 MB
localhost/ubi8/openjdk-11-runtime        1.3              370 MB
localhost/ubi8/openjdk-11                latest           474 MB
localhost/ubi8/openjdk-11                1.3              474 MB
localhost/ubi8/openjdk-8                 latest           479 MB
localhost/ubi8/openjdk-8                 1.3              479 MB
```

jmtd commented on modules/jre/run/bash/artifacts/opt/jboss/container/java/run:
```diff
@@ -0,0 +1,12 @@
+#!/bin/bash
+
+echo "Starting the Java application using ${JBOSS_CONTAINER_JAVA_RUN_MODULE}/run-java.sh ..."
```
Please remove this line

jmtd commented on modules/jre/run/bash/artifacts/opt/jboss/container/java/run-env.sh:
```diff
@@ -0,0 +1,4 @@
+#!/bin/sh
```
We don't need this file in the image, please remove it

jmtd commented on modules/jre/run/bash/artifacts/opt/jboss/container/java/run-java.sh:
```diff
@@ -0,0 +1,246 @@
+#!/bin/sh
```
We don't need this file in the image, please remove it

jmtd commented on modules/jre/run/bash/module.yaml:
```diff
@@ -0,0 +1,17 @@
+schema_version: 1
+name: jboss.container.java.jre.run.bash
```
Please rename the module to remove the ".bash" suffix, and move the files accordingly (from `modules/jre/run/bash` to `modules/jre/run`)

jmtd commented on ubi8-openjdk-11-runtime.yaml:
```diff
@@ -39,6 +39,7 @@ modules:
   install:
   - name: jboss.container.openjdk.jre
     version: "11"
+  - name: jboss.container.java.jre.run.bash
```
Once the module is renamed this will need adjusting

jmtd commented on ubi8-openjdk-8-runtime.yaml:
```diff
@@ -39,6 +39,7 @@ modules:
   install:
   - name: jboss.container.openjdk.jre
     version: "8"
+  - name: jboss.container.java.jre.run.bash
```
Once the module is renamed this will need adjusting

jmtd reviewed (changes requested):
This looks good - it's nearly there, just a few unneeded files and I'd prefer to avoid the ".bash" suffix on the module name. Thanks!

jhuttana commented:
@jmtd Thanks for the review. I have addressed the comments.

```
$ podman run --rm -ti localhost/ubi8/openjdk-11-runtime:latest
This is the Red Hat UBI8 OpenJDK 1.8.0 Runtime container.
$ podman run --rm -ti localhost/ubi8/openjdk-8-runtime:latest
This is the Red Hat UBI8 OpenJDK 1.8.0 Runtime container.
```

```$ podman images --filter dangling=false
REPOSITORY                              TAG      IMAGE ID      CREATED         SIZE
localhost/ubi8/openjdk-11-runtime     latest    686e1984d33a  14 seconds ago  370 MB
localhost/ubi8/openjdk-11-runtime     1.3       686e1984d33a  14 seconds ago  370 MB
localhost/ubi8/openjdk-8-runtime      latest    a789e9e7048c  49 seconds ago  314 MB
localhost/ubi8/openjdk-8-runtime      1.3       a789e9e7048c  49 seconds ago  314 MB
```

jmtd commented:
Hi @jhuttana, no problem! The PR now only modifies the two runtime image descriptors (to add jboss.container.java.jre.run to the list of modules) but does not contain the module sources. Please `git add` them.

jhuttana commented:
> Hi @jhuttana, no problem! The PR now only modifies the two runtime image descriptors (to add jboss.container.java.jre.run to the list of modules) but does not contain the module sources. Please `git add` them.

Ooops !! My bad , somehow missed it :) Now corrected. Could you please check?

jmtd reviewed (approved):
👍🏼

flybyray commented:
@jmtd is there any link to some design decissions for the runtime images. how you would use them in production?
the image should work on Kubernetes/openshift? the non-runtime images have some scripts which evaluate the resource limits set on containers and inject specific flags for java jvm. How this should be handled in the runtime images? is there some documentation about it?
I found for instance this blog https://developers.redhat.com/articles/2021/05/24/build-lean-java-containers-new-red-hat-universal-base-images-openjdk-runtime#a_runtime_image
which uses the runtime image via buildconfig. but actually the result will not be run with the same jvm settings ( depending on the buildconfig).
is it meant to have a buildconfig which tries to copy scripts and CMDs from the intermediate image?

jmtd commented:
Hi @flybyray, that blog post is currently the only documentation for the runtime images. You're absolutely right, that the runtime image won't have the same JVM settings as the full builder images and is not influenced or configured using the same environment variables. We made a decision to ship the first iteration of the runtime images without a run script, in order to get them out quickly, for two reasons:

1. reportedly a lot of users of the images completely override the JVM settings and do not use our environment variables to tweak them, instead providing a pre-baked static set of JVM settings (perhaps generated using some external generator tool). Those folks can easily supply such a static list in the Dockerfile deployment config workflow described in that post.
2. to get the initial release of the runtime images out quickly. The existing run-java script that is in the builder image is closely coupled with the s2i scripts and the agent JARs that are shipped in the builder images. We need to go through run-java carefully to establish what makes sense to include in the runtime image and refactor accordingly. There's some other work that needs to be done around run-java which would make sense to do at the same time as this,  but would have delayed getting an initial image out. We've got issue #220 filed to track this.
